pkg/querier/tripperware/shard_by_query.go

```diff
@@ -0,0 +1,81 @@
+package tripperware
+
+import (
+	"context"
+
+	"github.com/go-kit/log"
+	"github.com/thanos-io/thanos/pkg/querysharding"
+	"github.com/thanos-io/thanos/pkg/store/storepb"
+
+	cquerysharding "github.com/cortexproject/cortex/pkg/querysharding"
+)
+
+func ShardByMiddleware(logger log.Logger, limits Limits, merger Merger, numShards int) Middleware {
+	return MiddlewareFunc(func(next Handler) Handler {
+		return shardQueryByMiddleware{
+			next:      next,
+			limits:    limits,
+			merger:    merger,
+			logger:    logger,
+			numShards: numShards,
+		}
+	})
+}
+
+type shardQueryByMiddleware struct {
+	next          Handler
+	limits        Limits
+	logger        log.Logger
+	merger        Merger
+	queryAnalyzer *querysharding.QueryAnalyzer
+	numShards     int
+}
+
+func (s shardQueryByMiddleware) Do(ctx context.Context, r Request) (Response, error) {
+	analysis, err := s.queryAnalyzer.Analyze(r.GetQuery())
+	if err != nil {
```

Comment:
Maybe you also want to bring https://github.com/thanos-io/thanos/pull/5684 in?

Reply:
Good catch! Thanks!!!